OpenSim/Common/Logger.cpp

```diff
@@ -30,23 +30,93 @@
 
 using namespace OpenSim;
 
-std::shared_ptr<spdlog::logger> Logger::m_cout_logger = 
-        spdlog::stdout_color_mt("cout");
-std::shared_ptr<spdlog::sinks::basic_file_sink_mt> Logger::m_filesink = {};
-std::shared_ptr<spdlog::logger> Logger::m_default_logger;
-
-// Force creation of the Logger instane to initialize spdlog::loggers
-std::shared_ptr<OpenSim::Logger> Logger::m_osimLogger = Logger::getInstance();
-
-Logger::Logger() {
-    m_default_logger = spdlog::default_logger();
-    m_default_logger->set_level(spdlog::level::info);
-    m_default_logger->set_pattern("[%l] %v");
-    m_cout_logger->set_level(spdlog::level::info);
-    m_cout_logger->set_pattern("%v");
-    // This ensures log files are updated regularly, instead of only when the
-    // program shuts down.
+// cout logger will be initialized during static initialization time
+static std::shared_ptr<spdlog::logger> cout_logger = []() {
+    std::shared_ptr<spdlog::logger> logger = spdlog::stdout_color_mt("cout");
+    logger->set_level(spdlog::level::info);
+    logger->set_pattern("%v");
+    return logger;
+}();
+
+// default logger will be initialized during static initialization time
+static std::shared_ptr<spdlog::logger> default_logger = []() {
+    std::shared_ptr<spdlog::logger> logger = spdlog::default_logger();
```

Comment:
Just so I understand, what is the C++11 syntax for initializing a static variable exactly once at static initialization time, bar writing a separate function?:

```c++
T f() {
    // multiple init steps
}

static T = f();
```

(assuming the initialization has extra steps beyond the constructor)

Reply:
Just to clarify what I'm getting at - the existing way was *definitely not* the way to do it. C++ doesn't have static classes, or static constructors (unlike e.g. Java) and the idea of constructing an instance of the class to initialize static member variables was asking for an issue somewhere.

Entirely unrelated to this PR (purely coincidence), writing static init code like that is exactly was why we ended up with a memory leak:

https://github.com/opensim-org/opensim-core/issues/2865

The only way to guarantee one-time global initialization is by having a single initializing statement for each element being initialized, or to declare a non-static class with non-static members and *then* initialize a static instance of that class:

```c++
T f() { ... }
static T t = f();  // ok

class C { T m1 = f(); T m2 = f(); };
static C t;  // also ok

// same as first case, but don't expose `f` for use anywhere else:
static T t = []() { ... }();
```

Reply:
Thanks @adamkewley I was advocating for the first syntax despite the fact that it exposes the name unnecessarily  simply because the rest of the code is c++03 (my bad I claimed c++11) and I couldn't find any instance of the unnamed function initialization in our code base . Purely code style.

Reply:
Ah - got it. I could change it to use named functions if it's anoying enough (it's not a cross i'm willing to die on)

Reply:
Let's do it then. Big thanks for accommodating 👍